torch/csrc/distributed/rpc/tensorpipe_agent.cpp

```diff
@@ -133,36 +184,40 @@ TensorPipeAgent::~TensorPipeAgent() {
 
 void TensorPipeAgent::startImpl() {
   VLOG(1) << "RPC agent for " << workerInfo_.name_ << " is starting";
-  auto uvContext = std::make_shared<tensorpipe::transport::uv::Context>();
 
-  std::string uvAddress = guessAddress(*uvContext);
-  VLOG(1) << "RPC agent for " << workerInfo_.name_ << " is using address "
-          << uvAddress;
+  std::vector<std::string> addresses;
+  int lowestPriority = -1;
+  std::string lowestPriorityTransport;
 
-  context_->registerTransport(1, "tcp", std::move(uvContext));
-#ifdef TP_ENABLE_SHM
-  context_->registerTransport(
-      0, "shm", std::make_shared<tensorpipe::transport::shm::Context>());
-#endif
-  context_->registerChannel(
-      1, "basic", std::make_shared<tensorpipe::channel::basic::Context>());
-#ifdef TP_ENABLE_CMA
-  context_->registerChannel(
-      0, "cma", std::make_shared<tensorpipe::channel::cma::Context>());
-#endif
+  for (auto& key : TensorPipeTransportRegistry()->Keys()) {
+    std::unique_ptr<TransportRegistration> reg =
+        TensorPipeTransportRegistry()->Create(key);
+    if (reg->priority > lowestPriority) {
+      lowestPriority = reg->priority;
+      lowestPriorityTransport = key;
+    }
+    addresses.push_back(c10::str(key, "://", reg->address));
+    context_->registerTransport(
+        reg->priority, std::move(key), std::move(reg->transport));
+  }
 
-  std::vector<std::string> addresses = {"tcp://" + uvAddress};
-#ifdef TP_ENABLE_SHM
-  addresses.push_back(createUniqueShmAddr());
-#endif
+  for (auto& key : TensorPipeChannelRegistry()->Keys()) {
+    std::unique_ptr<ChannelRegistration> reg =
+        TensorPipeChannelRegistry()->Create(key);
+    context_->registerChannel(
+        reg->priority, std::move(key), std::move(reg->channel));
+  }
 
   listener_ = context_->listen(addresses);
 
   // Store our own url.
-  const auto address = listener_->url("tcp");
+  const auto address = listener_->url(lowestPriorityTransport);
   const std::vector<uint8_t> selfAddrData(address.begin(), address.end());
   nameToAddressStore_.set(workerInfo_.name_, selfAddrData);
```

Comment:
Curious about the future plan here. The current code looks like only registers the transport with the lowest priority to the name->address store. What about other transport? And what if this transport only connects this worker with a subset of the gang? How would the rest of the gang talk to this worker?

Reply:
So, each pipe has an "initial" transport, which is the one specified in the address used to connect (e.g., "uv" in `uv://127.0.0.1:1234` or "shm" in `shm://foobar`), which is the transport used to bootstrap the pipe. Afterwards, during the handshake, the pipe will try to negotiate a better transport, if it can find one. "Better" means a transport with a priority higher than the initial transport. "If it can find one" means a transport that both sides can use and that would work (e.g., SHM would not work if the two sides are on different machines). If such a transport is found the pipe switches over to that one, otherwise it remains on the initial one. (Note that the addresses for these other transports do not have to be stored in the nameToAddressStore, because the server will send them to the client over the pipe as part of the handshake).

What this means is that any transport that has a _lower_ priority than the initial transport will never be used (it makes no sense to downgrade to it, it will just make things worse). Therefore, the "converse" property, is that the initial transport should be the transport with lowest priority.

It also means that the transport with lowest priority should be (the only) one that is able to connect all workers. If that were not the case, then it means that there is another transport with higher priority (and thus better!) that works for all workers, but then we should make that transport the default and remove all the lower-priority ones.

All the above reasoning is based on the assumption that there is a strict total order of "goodness" among all transports, which is captured by the priority. This assumption works, for now, but is not always correct. You could imagine having both a TCP and an InfiniBand transport: the TCP one might provide a higher bandwidth, while the InfiniBand one might provide a lower latency. So there isn't one that is strictly better than the other, and thus we cannot rank them using priorities. This kind of more fine-grained selection is something we want to work on soon.